File: packages/wavesync-rn/ios/WaveSyncModule.swift
import Foundation

/// React Native native module for WaveSyncDB.
/// Mirrors the Android WaveSyncModule.kt — same method names, same JS API.
@objc(WaveSync)
class WaveSyncModule: RCTEventEmitter {

    private var ffi: WaveSyncFfi?
    private var hasListeners = false

    override init() {
        super.init()
    }

    @objc override static func requiresMainQueueSetup() -> Bool { false }

    override func supportedEvents() -> [String]! {
        ["WaveSyncChange", "WaveSyncNetworkEvent"]
    }

    override func startObserving() { hasListeners = true }
    override func stopObserving() { hasListeners = false }

    // MARK: - Initialization

    @objc func initialize(_ topic: String, options: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Task {
            do {
                let docsDir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
                let dbPath = "\(docsDir)/wavesync.db"
                let dbUrl = "sqlite://\(dbPath)?mode=rwc"

                var builder = WaveSyncFfiBuilder(dbUrl: dbUrl, topic: topic)

                if let passphrase = options["passphrase"] as? String {
                    builder = builder.withPassphrase(passphrase: passphrase)
                }
                if let relay = options["relayServer"] as? String {
                    builder = builder.withRelayServer(addr: relay)
                }
                if let rendezvous = options["rendezvousServer"] as? String {
                    builder = builder.withRendezvousServer(addr: rendezvous)
                }
                if let bootstrap = options["bootstrapPeer"] as? String {
                    builder = builder.withBootstrapPeer(addr: bootstrap)
                }
                if let addr = options["managedRelayAddr"] as? String,
                   let key = options["managedRelayApiKey"] as? String {
                    builder = builder.managedRelay(addr: addr, apiKey: key)
                }
                if let ipv6 = options["ipv6"] as? Bool {
                    builder = builder.withIpv6(enabled: ipv6)
                }
                if let interval = options["syncIntervalSeconds"] as? NSNumber {
                    builder = builder.withSyncInterval(seconds: interval.uint64Value)
                }
                if let keepAlive = options["keepAliveIntervalSeconds"] as? NSNumber {
                    builder = builder.withKeepAliveInterval(seconds: keepAlive.uint64Value)
                }

                self.ffi = try await builder.build()
                resolve(true)
            } catch {
                reject("WAVESYNC_ERR", error.localizedDescription, error)
            }
        }
    }

    // MARK: - SQL Operations

    @objc func execute(_ sql: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        Task {
            do {
                let rows = try await engine.execute(sql: sql)
                resolve(NSNumber(value: rows))
            } catch {
                reject("WAVESYNC_ERR", error.localizedDescription, error)
            }
        }
    }

    @objc func query(_ sql: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        Task {
            do {
                let json = try await engine.query(sql: sql)
                resolve(json)
            } catch {
                reject("WAVESYNC_ERR", error.localizedDescription, error)
            }
        }
    }

    // MARK: - Table Registration

    @objc func registerSyncedTable(_ tableName: String, pkColumn: String, columns: [String], createSql: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        Task {
            do {
                try await engine.registerSyncedTable(tableName: tableName, pkColumn: pkColumn, columns: columns, createSql: createSql)
                resolve(true)
            } catch {
                reject("WAVESYNC_ERR", error.localizedDescription, error)
            }
        }
    }

    @objc func registryReady(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.registryReady()
        resolve(true)
    }

    // MARK: - Network Status

    @objc func networkStatus(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        let status = engine.networkStatus()
        let peers = status.peers.map { peer -> [String: Any] in
            var map: [String: Any] = [
                "peerId": peer.peerId,
                "address": peer.address,
                "isBootstrap": peer.isBootstrap,
                "isGroupMember": peer.isGroupMember,
            ]
            if let v = peer.dbVersion { map["dbVersion"] = NSNumber(value: v) }
            if let id = peer.appId { map["appId"] = id }
            return map
        }
        let map: [String: Any] = [
            "localPeerId": status.localPeerId,
            "peerCount": status.peerCount,
            "groupPeerCount": status.groupPeerCount,
            "relayStatus": String(describing: status.relayStatus).uppercased(),
            "natStatus": String(describing: status.natStatus).uppercased(),
            "topic": status.topic,
            "rendezvousRegistered": status.rendezvousRegistered,
            "pushRegistered": status.pushRegistered,
            "localDbVersion": NSNumber(value: status.localDbVersion),
            "registryReady": status.registryReady,
            "peers": peers,
        ]
        resolve(map)
    }

    // MARK: - Subscriptions

    @objc func subscribeChanges(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.subscribeChanges(listener: ChangeListenerImpl(module: self))
        resolve(true)
    }

    @objc func subscribeNetworkEvents(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.subscribeNetworkEvents(listener: NetworkEventListenerImpl(module: self))
        resolve(true)
    }

    // MARK: - Lifecycle & Sync Control

    @objc func resume(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.resume()
        resolve(true)
    }

    @objc func networkTransition(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.networkTransition()
        resolve(true)
    }

    @objc func requestFullSync(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.requestFullSync()
        resolve(true)
    }

    @objc func isEngineAlive(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        resolve(engine.isEngineAlive())
    }

    @objc func registerPushToken(_ platform: String, token: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.registerPushToken(platform: platform, token: token)
        resolve(true)
    }

    // MARK: - Peer Identity

    @objc func setPeerIdentity(_ appId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.setPeerIdentity(appId: appId)
        resolve(true)
    }

    @objc func clearPeerIdentity(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        engine.clearPeerIdentity()
        resolve(true)
    }

    // MARK: - Database Metadata

    @objc func nodeId(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        resolve(engine.nodeId())
    }

    @objc func siteId(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        resolve(engine.siteId())
    }

    @objc func readSyncConfig(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return reject("NOT_INIT", "Call initialize first", nil) }
        do {
            resolve(try engine.readSyncConfig())
        } catch {
            reject("WAVESYNC_ERR", error.localizedDescription, error)
        }
    }

    // MARK: - Shutdown

    @objc func shutdown(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let engine = ffi else { return resolve(true) }
        Task {
            do {
                try await engine.shutdown()
                self.ffi = nil
                resolve(true)
            } catch {
                reject("WAVESYNC_ERR", error.localizedDescription, error)
            }
        }
    }

    // MARK: - Event Emitter Stubs

    @objc func addListener(_ eventName: String) { /* required by NativeEventEmitter */ }
    @objc func removeListeners(_ count: Double) { /* required by NativeEventEmitter */ }

    // MARK: - Internal: Send events to JS

    func emitChange(_ notification: FfiChangeNotification) {
        guard hasListeners else { return }
        var params: [String: Any] = [
            "table": notification.table,
            "kind": String(describing: notification.kind).uppercased(),
            "primaryKey": notification.primaryKey,
        ]
        if let cols = notification.changedColumns {
            params["changedColumns"] = cols
        }
        sendEvent(withName: "WaveSyncChange", body: params)
    }

    func emitNetworkEvent(_ event: FfiNetworkEvent) {
        guard hasListeners else { return }
        var params: [String: Any] = [:]
        switch event {
        case .peerConnected(let peer):
            params["type"] = "PeerConnected"
            params["peer"] = peerInfoToDict(peer)
        case .peerDisconnected(let peerId):
            params["type"] = "PeerDisconnected"
            params["peerId"] = peerId
        case .peerRejected(let peerId):
            params["type"] = "PeerRejected"
            params["peerId"] = peerId
        case .peerVerified(let peerId):
            params["type"] = "PeerVerified"
            params["peerId"] = peerId
        case .peerIdentityReceived(let peerId, let appId):
            params["type"] = "PeerIdentityReceived"
            params["peerId"] = peerId
            params["appId"] = appId
        case .relayStatusChanged(let status):
            params["type"] = "RelayStatusChanged"
            params["status"] = String(describing: status).uppercased()
        case .natStatusChanged(let status):
            params["type"] = "NatStatusChanged"
            params["status"] = String(describing: status).uppercased()
        case .rendezvousStatusChanged(let registered):
            params["type"] = "RendezvousStatusChanged"
            params["registered"] = registered
        case .peerSynced(let peerId, let dbVersion):
            params["type"] = "PeerSynced"
            params["peerId"] = peerId
            params["dbVersion"] = NSNumber(value: dbVersion)
        case .engineStarted:
            params["type"] = "EngineStarted"
        case .engineFailed(let reason):
            params["type"] = "EngineFailed"
            params["reason"] = reason
        }
        sendEvent(withName: "WaveSyncNetworkEvent", body: params)
    }

    private func peerInfoToDict(_ peer: FfiPeerInfo) -> [String: Any] {
        var map: [String: Any] = [
            "peerId": peer.peerId,
            "address": peer.address,
            "isBootstrap": peer.isBootstrap,
            "isGroupMember": peer.isGroupMember,
        ]
        if let v = peer.dbVersion { map["dbVersion"] = NSNumber(value: v) }
        if let id = peer.appId { map["appId"] = id }
        return map
    }
}

// MARK: - Callback Implementations

class ChangeListenerImpl: ChangeListener {
    private weak var module: WaveSyncModule?

    init(module: WaveSyncModule) {
        self.module = module
    }

    func onChange(notification: FfiChangeNotification) {
        module?.emitChange(notification)
    }
}

class NetworkEventListenerImpl: NetworkEventListener {
    private weak var module: WaveSyncModule?

    init(module: WaveSyncModule) {
        self.module = module
    }

    func onEvent(event: FfiNetworkEvent) {
        module?.emitNetworkEvent(event)
    }
}

File: packages/wavesync-rn/android/src/main/java/com/wavesync/WaveSyncModule.kt
package com.wavesync

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import kotlinx.coroutines.*
import uniffi.wavesyncdb_ffi.*

class WaveSyncModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "WaveSync"

    private var ffi: WaveSyncFfi? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // -----------------------------------------------------------------------
    // Initialization
    // -----------------------------------------------------------------------

    @ReactMethod
    fun initialize(topic: String, options: ReadableMap, promise: Promise) {
        scope.launch {
            try {
                val dbPath = "${reactApplicationContext.filesDir.absolutePath}/wavesync.db"
                val dbUrl = "sqlite://$dbPath?mode=rwc"

                var builder = WaveSyncFfiBuilder(dbUrl, topic)

                options.getString("passphrase")?.let {
                    builder = builder.withPassphrase(it)
                }
                options.getString("relayServer")?.let {
                    builder = builder.withRelayServer(it)
                }
                options.getString("rendezvousServer")?.let {
                    builder = builder.withRendezvousServer(it)
                }
                options.getString("bootstrapPeer")?.let {
                    builder = builder.withBootstrapPeer(it)
                }
                options.getString("managedRelayAddr")?.let { addr ->
                    options.getString("managedRelayApiKey")?.let { key ->
                        builder = builder.managedRelay(addr, key)
                    }
                }
                if (options.hasKey("ipv6")) {
                    builder = builder.withIpv6(options.getBoolean("ipv6"))
                }
                if (options.hasKey("syncIntervalSeconds")) {
                    builder = builder.withSyncInterval(options.getDouble("syncIntervalSeconds").toLong().toULong())
                }
                if (options.hasKey("keepAliveIntervalSeconds")) {
                    builder = builder.withKeepAliveInterval(options.getDouble("keepAliveIntervalSeconds").toLong().toULong())
                }

                ffi = builder.build()
                promise.resolve(true)
            } catch (e: WaveSyncException) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            } catch (e: Exception) {
                promise.reject("WAVESYNC_INIT_ERR", e.message, e)
            }
        }
    }

    // -----------------------------------------------------------------------
    // SQL operations
    // -----------------------------------------------------------------------

    @ReactMethod
    fun execute(sql: String, promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        scope.launch {
            try {
                val rowsAffected = engine.execute(sql)
                promise.resolve(rowsAffected.toLong().toDouble())
            } catch (e: WaveSyncException) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun query(sql: String, promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        scope.launch {
            try {
                val jsonStr = engine.query(sql)
                promise.resolve(jsonStr)
            } catch (e: WaveSyncException) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            }
        }
    }

    // -----------------------------------------------------------------------
    // Table registration
    // -----------------------------------------------------------------------

    @ReactMethod
    fun registerSyncedTable(
        tableName: String,
        pkColumn: String,
        columns: ReadableArray,
        createSql: String,
        promise: Promise
    ) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        scope.launch {
            try {
                val colList = (0 until columns.size()).map { columns.getString(it)!! }
                engine.registerSyncedTable(tableName, pkColumn, colList, createSql)
                promise.resolve(true)
            } catch (e: WaveSyncException) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun registryReady(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        try {
            engine.registryReady()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("WAVESYNC_ERR", e.message, e)
        }
    }

    // -----------------------------------------------------------------------
    // Network status
    // -----------------------------------------------------------------------

    @ReactMethod
    fun networkStatus(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        try {
            val status = engine.networkStatus()
            val map = Arguments.createMap().apply {
                putString("localPeerId", status.localPeerId)
                putInt("peerCount", status.peerCount.toInt())
                putInt("groupPeerCount", status.groupPeerCount.toInt())
                putString("relayStatus", status.relayStatus.name)
                putString("natStatus", status.natStatus.name)
                putString("topic", status.topic)
                putBoolean("rendezvousRegistered", status.rendezvousRegistered)
                putBoolean("pushRegistered", status.pushRegistered)
                putDouble("localDbVersion", status.localDbVersion.toLong().toDouble())
                putBoolean("registryReady", status.registryReady)

                val peersArray = Arguments.createArray()
                for (peer in status.peers) {
                    peersArray.pushMap(peerInfoToMap(peer))
                }
                putArray("peers", peersArray)
            }
            promise.resolve(map)
        } catch (e: Exception) {
            promise.reject("WAVESYNC_ERR", e.message, e)
        }
    }

    // -----------------------------------------------------------------------
    // Subscriptions
    // -----------------------------------------------------------------------

    @ReactMethod
    fun subscribeChanges(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.subscribeChanges(object : ChangeListener {
            override fun onChange(notification: FfiChangeNotification) {
                try {
                    val params = Arguments.createMap().apply {
                        putString("table", notification.table)
                        putString("kind", notification.kind.name)
                        putString("primaryKey", notification.primaryKey)
                        notification.changedColumns?.let { cols ->
                            val arr = Arguments.createArray()
                            cols.forEach { arr.pushString(it) }
                            putArray("changedColumns", arr)
                        }
                    }
                    reactApplicationContext
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                        .emit("WaveSyncChange", params)
                } catch (_: Exception) {
                    // Context may be destroyed; silently ignore
                }
            }
        })
        promise.resolve(true)
    }

    @ReactMethod
    fun subscribeNetworkEvents(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.subscribeNetworkEvents(object : NetworkEventListener {
            override fun onEvent(event: FfiNetworkEvent) {
                try {
                    val params = Arguments.createMap().apply {
                        when (event) {
                            is FfiNetworkEvent.PeerConnected -> {
                                putString("type", "PeerConnected")
                                putMap("peer", peerInfoToMap(event.peer))
                            }
                            is FfiNetworkEvent.PeerDisconnected -> {
                                putString("type", "PeerDisconnected")
                                putString("peerId", event.peerId)
                            }
                            is FfiNetworkEvent.PeerRejected -> {
                                putString("type", "PeerRejected")
                                putString("peerId", event.peerId)
                            }
                            is FfiNetworkEvent.PeerVerified -> {
                                putString("type", "PeerVerified")
                                putString("peerId", event.peerId)
                            }
                            is FfiNetworkEvent.PeerIdentityReceived -> {
                                putString("type", "PeerIdentityReceived")
                                putString("peerId", event.peerId)
                                putString("appId", event.appId)
                            }
                            is FfiNetworkEvent.RelayStatusChanged -> {
                                putString("type", "RelayStatusChanged")
                                putString("status", event.status.name)
                            }
                            is FfiNetworkEvent.NatStatusChanged -> {
                                putString("type", "NatStatusChanged")
                                putString("status", event.status.name)
                            }
                            is FfiNetworkEvent.RendezvousStatusChanged -> {
                                putString("type", "RendezvousStatusChanged")
                                putBoolean("registered", event.registered)
                            }
                            is FfiNetworkEvent.PeerSynced -> {
                                putString("type", "PeerSynced")
                                putString("peerId", event.peerId)
                                putDouble("dbVersion", event.dbVersion.toLong().toDouble())
                            }
                            is FfiNetworkEvent.EngineStarted -> {
                                putString("type", "EngineStarted")
                            }
                            is FfiNetworkEvent.EngineFailed -> {
                                putString("type", "EngineFailed")
                                putString("reason", event.reason)
                            }
                        }
                    }
                    reactApplicationContext
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                        .emit("WaveSyncNetworkEvent", params)
                } catch (_: Exception) {
                    // Context may be destroyed; silently ignore
                }
            }
        })
        promise.resolve(true)
    }

    // -----------------------------------------------------------------------
    // Lifecycle & sync control
    // -----------------------------------------------------------------------

    @ReactMethod
    fun resume(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.resume()
        promise.resolve(true)
    }

    @ReactMethod
    fun networkTransition(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.networkTransition()
        promise.resolve(true)
    }

    @ReactMethod
    fun requestFullSync(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.requestFullSync()
        promise.resolve(true)
    }

    @ReactMethod
    fun isEngineAlive(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        promise.resolve(engine.isEngineAlive())
    }

    @ReactMethod
    fun registerPushToken(platform: String, token: String, promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.registerPushToken(platform, token)
        promise.resolve(true)
    }

    // -----------------------------------------------------------------------
    // Peer identity
    // -----------------------------------------------------------------------

    @ReactMethod
    fun setPeerIdentity(appId: String, promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.setPeerIdentity(appId)
        promise.resolve(true)
    }

    @ReactMethod
    fun clearPeerIdentity(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        engine.clearPeerIdentity()
        promise.resolve(true)
    }

    // -----------------------------------------------------------------------
    // Database metadata
    // -----------------------------------------------------------------------

    @ReactMethod
    fun nodeId(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        promise.resolve(engine.nodeId())
    }

    @ReactMethod
    fun siteId(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        promise.resolve(engine.siteId())
    }

    @ReactMethod
    fun readSyncConfig(promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        try {
            promise.resolve(engine.readSyncConfig())
        } catch (e: WaveSyncException) {
            promise.reject("WAVESYNC_ERR", e.message, e)
        }
    }

    @ReactMethod
    fun testBackgroundSync(timeoutSecs: Double, promise: Promise) {
        val engine = ffi ?: return promise.reject("NOT_INIT", "Call initialize first")
        val dbUrl = engine.readSyncConfig().let { json ->
            Regex(""""database_url"\s*:\s*"([^"]+)"""").find(json)?.groupValues?.get(1)
        } ?: return promise.reject("WAVESYNC_ERR", "Cannot extract database_url from config")

        scope.launch {
            try {
                // Shutdown the current engine first — background_sync creates its own
                engine.shutdown()
                ffi = null

                val result = uniffi.wavesyncdb_ffi.backgroundSync(
                    dbUrl,
                    timeoutSecs.toInt().toUInt(),
                    null
                )
                val map = Arguments.createMap().apply {
                    when (result) {
                        is FfiBackgroundSyncResult.Synced -> {
                            putString("status", "synced")
                            putInt("peersSynced", result.peersSynced.toInt())
                        }
                        is FfiBackgroundSyncResult.NoPeers -> {
                            putString("status", "no_peers")
                            putInt("peersSynced", 0)
                        }
                        is FfiBackgroundSyncResult.TimedOut -> {
                            putString("status", "timed_out")
                            putInt("peersSynced", result.peersSynced.toInt())
                        }
                    }
                }
                promise.resolve(map)
            } catch (e: Exception) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            }
        }
    }

    // -----------------------------------------------------------------------
    // Shutdown
    // -----------------------------------------------------------------------

    @ReactMethod
    fun shutdown(promise: Promise) {
        val engine = ffi ?: return promise.resolve(true)
        scope.launch {
            try {
                engine.shutdown()
                ffi = null
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("WAVESYNC_ERR", e.message, e)
            }
        }
    }

    // Required stubs for NativeEventEmitter
    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Double) {}

    @Deprecated("Deprecated in ReactContextBaseJavaModule")
    override fun onCatalystInstanceDestroy() {
        val engine = ffi ?: return
        ffi = null
        // Await engine shutdown so the SQLite connection is released before
        // the next initialize() call.  Without this, hot-reload or app restart
        // can hit "database is locked" because the old engine is still running.
        runBlocking {
            withTimeoutOrNull(5000) {
                engine.shutdown()
            }
        }
        scope.cancel()
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private fun peerInfoToMap(peer: FfiPeerInfo): WritableMap {
        return Arguments.createMap().apply {
            putString("peerId", peer.peerId)
            putString("address", peer.address)
            peer.dbVersion?.let {
                putDouble("dbVersion", it.toLong().toDouble())
            }
            putBoolean("isBootstrap", peer.isBootstrap)
            putBoolean("isGroupMember", peer.isGroupMember)
            peer.appId?.let { putString("appId", it) }
        }
    }
}
